### PR TITLE
Add budget component

### PR DESCRIPTION
**Pull request by andrewleith**
# Summary | Résumé

This PR adds the new "budget" component to Notify.  [Designs are here
](https://www.figma.com/design/JGp5noL1kCpiwJzHKL5MHe/Sending-capacity---capacit%C3%A9-d'envoi-(Copy)?node-id=2831-6844&node-type=canvas&focus-id=3083-819&m=dev)

## Component description
This component displays the number of remaining notifications (either SMS or email) that the user has left for both the day and the year.  The component uses a status indicator to convey when users are getting close to their limits (80%).   When a service gets to their limit, additional messages are displayed.  The component can also display using a text-only and a text-only with emoji mode.

### Below limits
image

### Near limits
image

### At daily limit
image

### At both limits
image

### Text-only
image

### Text-only with emojis
image

### CODE REVIEW

File: app/main/views/storybook.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from flask import render_template, request
+
+from app.main import main
+
+
+@main.route("/_storybook")
```
Comment (andrewleith) [inline app/main/views/storybook.py:6]: We need a new WAF rule to allow this to go through.  I'll create a rule to open it only for staging.

File: app/templates/components/remaining-messages-summary.html [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    {% set textOnlyMode = false %}
+  {% else %}
+    {% set textOnlyMode = true %}
+  {% endif %}
+
+  <!-- Validate notification_type param  -->
+  {% set notificationType_allowed_values = ['sms', 'email'] %}
+  {% if notification_type not in notificationType_allowed_values %}
+      {% set notification_type = None %}
+      {% if config["NOTIFY_ENVIRONMENT"].lower() == 'development' %}
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply app/templates/components/remaining-messages-summary.html:15]: That would be pretty cumbersome.  I think this is fine.

File: app/templates/components/remaining-messages-summary.html [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    {% if notification_type == 'sms' %}
+      {% set notification_type = _('text messages') %}
+    {% elif notification_type == 'email' %}
+      {% set notification_type = _('emails') %}
+    {% endif %}
+    {% set notification_type = notification_type %}
+  {% endif %}
+
+  <!-- Set some constants -->
+  {% set current_year = current_year or (now().year if now().month < 4 else now().year + 1) %}
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply app/templates/components/remaining-messages-summary.html:28]: This comes from the [jinja globals](https://github.com/cds-snc/notification-admin/pull/1979/files/8eaeeb6e35fd9b019bdc0a110df7a6b4c98c8542#diff-9cec7b11237bc29d77a439e81c9b7acfac003d8e8855731eb6bc130b5a8ce602)

File: app/templates/components/remaining-messages-summary.html [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  {% set sections = [
+    {
+      'type': 'daily',
+      'used': dailyUsed,
+      'limit': dailyLimit,
+      'text': _('remaining until') ~ ' ' ~ ('7 pm ET'),
+      'link_text': _('Request a daily limit increase'),
+      'link_href': url_for('main.contact'),
+      'remaining': "{:,}".format(dailyLimit - dailyUsed) if session['userlang'] == 'en' else "{:,}".format(dailyLimit - dailyUsed).replace(',', ' '),
+      'skip': false if textOnlyMode else dailyLimit - dailyUsed == 0 and yearlyLimit - yearlyUsed == 0,
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply app/templates/components/remaining-messages-summary.html:43]: Not worth the effort!

File: tests_cypress/cypress/e2e/admin/components/remaining_messages_summary.cy.js [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+      RMS.Below()
+        .find('*[data-testid="rms"]')
+        .eq(1)
+        .find('*[data-testid="rms-yearly-remaining"]')
+        .should("have.text", "9,301");
+    });
+
+    it("shows thousands separator in FR for the year’s remaining", () => {
+      cy.get("#header-lang").click();
+      cy.visit(PageURL);
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply tests_cypress/cypress/e2e/admin/components/remaining_messages_summary.cy.js:85]: No thanks!

File: tests_cypress/cypress/e2e/admin/components/remaining_messages_summary.cy.js [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+      RMS.TextWarn()
+        .find('*[data-testid="rms-item"]')
+        .eq(1)
+        .find('*[data-testid="text-prefix-near-limit"]')
+        .should("exist");
+    });
+  });
+
+  describe("Text-only at limit", () => {
+    it("shows “At limit” daily prefix when daily is < 80% usage", () => {
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply tests_cypress/cypress/e2e/admin/components/remaining_messages_summary.cy.js:191]: Good catch!

File: tests_cypress/cypress/e2e/admin/components/remaining_messages_summary.cy.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewleith) [reply tests_cypress/cypress/e2e/admin/components/remaining_messages_summary.cy.js:85]: dr-evil-how-about-no

File: app/templates/components/remaining-messages-summary.html [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            <span data-testid="rms-icon-{{ icon_type }}">
+              <i aria-hidden="true" class="fa-solid fa-fas {{ icon_class }} text-xs rounded-full mx-1" {{ 'aria-label=' ~ _('warning') if icon_type == "warning" else "" }}></i>
+            </span>
+            <span class="mx-2 font-bold"  data-testid="rms-{{ section.type }}-remaining">{{ section.remaining }}</span> {{ section.text }}
+          </div>
+          <a href="{{ section.link_href }}" data-testid="rms-{{ section.type }}-link" class="text-blue-500">{{ section.link_text }}</a>
+        </div>
+      {% else %}
+        <p class="m-0 p-0" data-testid="rms-item">
+          {% if section.remaining == "0" %}
```
[the remark this answers is not in the document]
Comment (andrewleith) [reply app/templates/components/remaining-messages-summary.html:80]: It is being treated as a string, hence the string compare.

File: app/templates/views/storybook/remaining-messages-summary.html [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    <div class="grid grid-cols-2 gap-6 bg-gray p-6" data-testid="rms-emoji-warning">near limit
+        <div class="mb-6 bg-white p-4">
+            {{ remaining_messages_summary(1000, 800, 1000, 900, "email", "emoji") }}
+        </div>
+    </div>
+    <div class="grid grid-cols-2 gap-6 bg-gray p-6" data-testid="rms-emoji-limit">at limit
+        <div class="mb-6 bg-white p-4">
+            {{ remaining_messages_summary(1000, 1000, 1000, 1000, "email", "emoji") }}
+        </div>
+    </div>
```
Comment (andrewleith) [inline app/templates/views/storybook/remaining-messages-summary.html:85]:
```suggestion
    </div>

```

File: app/main/views/storybook.py (hunk 1 above)
Comment (amazingphilippe) [reply app/main/views/storybook.py:6]: This is really cool! I can see the limit component, but the budget component throws an error. Should it be working on the review app? Or am I better to run it locally?

Comment (andrewleith) [reply app/main/views/storybook.py:6]: Ah it seems I listened to the AI review and implemented something that doesnt work 😮

Comment (andrewleith) [reply app/main/views/storybook.py:6]: Fixed now!

**Review by jzbahrai — APPROVED** [review]

**Pull request merged**